File: lib/types/graph.ex
defmodule Lego.Graph do
  use TypedStruct

  typedstruct do
    field :blocks, [Lego.Block.t()], default: []
  end

  def add_block(graph, block) do
    %{ graph | blocks: [block | graph.blocks] }
  end
end

File: src/graph.zig
const std = @import("std");
const Io = std.Io;
const ArrayList = std.ArrayList;
const Allocator = std.mem.Allocator;

const lego = @import("root.zig");
const Block = lego.Block;

pub const Graph = struct {
    allocator: Allocator,
    blocks: ArrayList(Block),

    pub fn init(gpa: Allocator) Graph {
        return .{
            .allocator = gpa,
            .blocks = .empty,
        };
    }

    pub fn addBlock(self: *Graph, block: Block) !void {
        try self.blocks.append(self.allocator, block);
    }
};
